### DISCUSSION TITLE
Realtime 401 error in Nextjs

### DISCUSSION THREAD

**Discussion by pablojsx** [supabase/supabase#26932] [Questions] [answered]
I'm starting to think this is a bug because no matter what I do I always get a 401 error. I already have a setup for uploading images to their bucket and works fine, tried to use the same supabase instance and didn't work. I keep reading and I found about the ssr client, cool, no luck. 

I have tried every youtube video, docs, no luck. I see a guy in stackoverflow having the same issue, it just return 401 all the time, RLS enabled with all policies combinations, disabled, no luck. 

Is this a bug?

![image](https://github.com/supabase/supabase/assets/40902487/bd10614f-56b5-415a-afe9-80ed3de484ef)

![image](https://github.com/supabase/supabase/assets/40902487/08cd7ecf-20fa-454c-9553-25e83ea79bba)

![image](https://github.com/supabase/supabase/assets/40902487/788d0c05-b648-4aa8-9a63-9c06b7eda6f7)

![image](https://github.com/supabase/supabase/assets/40902487/78fe8840-e0c8-477e-a750-59cb0633f40d)

**Comment by pablojsx** [comment]
All combinations, around 6hrs, all I get is "Error 401: Unauthorized"

**Reply by hammer0x419** [reply]
have same issue, you are not crazy

**Comment by pablojsx** [comment]
It's been 9 days of emails and supabase still won't fix this issue, which at this point comes from them since I've done everything available and it's dumb because it isn't even hard to make this work but for me it's just 401.

**Comment by pablojsx** [comment]
Dear diary, supabase sent me an email suggesting the use of a policy I used in day 1 .... it's been 2 weeks of nonsense.

._.

**Comment by ladderschool** [comment]
Same thing!

**Comment by GaryAustin1** [comment]
401 can be grants are missing on the schema or tables.  This usually comes up if something like Prisma is used.  Prisma during migrations can delete the grants so the API no longer has access.    This would also though mean any select operations from next.js would not work.

A policy of `anon` also means an authenticated user will not have access.  But with row level security disabled the RLS is not involved.

**Reply by pablojsx** [reply]
Hello Gary, can you give more details about the grants? I'm using Prisma and Nextjs. I also get 401 with the Supabase inspector.

**Reply by GaryAustin1** [reply, marked as answer]
See the trouble shooting section here for some code to fix the grants if that is your issue.
https://supabase.com/partners/integrations/prisma
Prisma during migrations often wipes out Postgres grants to the public schema.
```
grant usage on schema public to postgres, anon, authenticated, service_role;

grant all privileges on all tables in schema public to postgres, anon, authenticated, service_role;
grant all privileges on all functions in schema public to postgres, anon, authenticated, service_role;
grant all privileges on all sequences in schema public to postgres, anon, authenticated, service_role;

alter default privileges in schema public grant all on tables to postgres, anon, authenticated, service_role;
alter default privileges in schema public grant all on functions to postgres, anon, authenticated, service_role;
alter default privileges in schema public grant all on sequences to postgres, anon, authenticated, service_role;
```

**Reply by ladderschool** [reply]
That fixed it for me, that's exactly what it was since I was using Prisma. Thank you!

**Reply by mark-antal-csizmadia** [reply]
This worked, thanks!

**Reply by stpn48** [reply]
Oh wow, thanks bro worked for me ❤️. Just copy paste those commands into supabase SQL editor for anyone wondering.

**Reply by bySharkster** [reply]
> Thanks God... more thank 8 hours on this rabbit hole

**Reply by hammer0x419** [reply]
Praise Gary. Gary, I just want to say that on this day, you saved my sanity. Thank you.

**Reply by hacking-impossible** [reply]
Wonderful! I was stuck on this issue over 4 hours and Claude, GPT couldn't solved this issue as well.

@GaryAustin1 Your experience surely helpful for me!

God bless you!

**Reply by elie222** [reply]
Ya. Also wasted endless time. The prisma supabase used to have more info about this, but seems like it got removed.

**Reply by xSenny** [reply]
Now if I use this code, it won't send realtime webhooks anymore. Only after revoking all those permissions

**Reply by jazib450** [reply]
> Now if I use this code, it won't send realtime webhooks anymore. Only after revoking all those permissions

@xSenny Same thing for me too. Did you end up figuring out the real problem?


EDIT:

What fixed it for me was putting it inside an async function and adding an await (any await seems to work) before the subscribe. It seems like it was subscribing before the session JWT was ready so it was subscribing unauthenticated. Something like the following:

```
useEffect(() => {
  const supabase = supabaseRef.current!;
  if (!ownerId) return;

  let channel: ReturnType<typeof supabase.channel> | null = null;
  let cancelled = false;

  (async () => {
    const { data: { session } } = await supabase.auth.getSession(); // 👈 hydrate token
    if (!session || cancelled) return;

    channel = supabase
      .channel(`conv:list:owner:${ownerId}`)
      .on(
        "postgres_changes",
        { event: "UPDATE", schema: "public", table: "Conversation", filter: `ownerId=eq.${ownerId}` },
        (payload) => {
          // do stuff
        }
      )
      .subscribe();
  })();

  return () => {
    cancelled = true;
    if (channel) supabase.removeChannel(channel);
  };
}, [ownerId]);
```
